### PR TITLE
feat: support one-off percentage overrides for user operations

### PR DESCRIPTION
**Pull request by denniswon**
## PR-Codex overview
This PR focuses on improving the fee options and overrides functionality in the code. 

### Detailed summary:
- Renamed `applyFeeOption` to `applyUserOpFeeOption` and added new functions `applyUserOpOverride` and `applyUserOpOverrideOrFeeOption`.
- Updated the `UserOperationOverrides` type to include new fields for overrides and fee options.
- Updated the `withFeeDataGetter` method in the provider to handle the new overrides and fee options.
- Updated the gas manager middleware to use the new functions for applying overrides and fee options.
- Updated the gas fees middleware to use the new functions for applying overrides and fee options.
- Updated the base provider to use the new functions for applying overrides and fee options.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/alchemy/src/middleware/gas-fees.ts [hunk 1]
```diff
@@ -1,22 +1,28 @@
...
+      : // it's a fair assumption that if someone is using this Alchemy Middleware, then they are using Alchemy RPC
+        applyFeeOption(
+          await (provider.rpcClient as ClientWithAlchemyMethods).request({
+            method: "rundler_maxPriorityFeePerGas",
+            params: [],
+          }),
+          isPercentage(overrides?.maxPriorityFeePerGas)
+            ? overrides?.maxPriorityFeePerGas
+            : feeOptions?.maxPriorityFeePerGas
+        );
```
Comment (moldy530) [inline packages/alchemy/src/middleware/gas-fees.ts:25]: can we not use the `applyOverride` method to handle this logic here?

File: packages/core/src/provider/base.ts [hunk 2]
```diff
@@ -555,33 +557,44 @@ export class SmartAccountProvider<
     let { callGasLimit, verificationGasLimit, preVerificationGas } =
       overrides ?? {};
 
+    const bn = "0x345345";
+    isBigNumberish(bn);
```
Comment (avasisht23) [inline packages/core/src/provider/base.ts:561]: remove testing?
```suggestion
```

File: packages/alchemy/src/middleware/gas-fees.ts (hunk 1 above)
Comment (denniswon) [reply packages/alchemy/src/middleware/gas-fees.ts:25]: hmm okay. but any critical reason?

Comment (denniswon) [reply packages/alchemy/src/middleware/gas-fees.ts]: updated

**Review by moldy530 — APPROVED** [review]

File: packages/core/src/provider/base.ts [hunk 3]
```diff
@@ -521,41 +522,39 @@ export class SmartAccountProvider<
...
-    ) {
-      const request = deepHexlify(await resolveProperties(struct));
-      const estimates = await this.rpcClient.estimateUserOperationGas(
-        request,
-        this.getEntryPointAddress()
+    const callGasLimit =
+      applyUserOperationOverride(
+        estimates.callGasLimit,
+        overrides?.callGasLimit
+      ) || applyFeeOption(estimates.callGasLimit, feeOptions?.callGasLimit);
```
Comment (avasisht23) [inline packages/core/src/provider/base.ts:535]: can you wrap this in a util? seems repeated below but generalizable on input

**Review by avasisht23 — APPROVED** [review]
nits, lgtm

**Comment by denniswon** [comment]
@avasisht23 accommodated your comment. thank you!

**Pull request merged**